### ISSUE TITLE
Compute agrob zed-velodyne cost function

### ISSUE THREAD

**Issue by aaguiar96** [lardemua/atom#149]

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 
I've been busy these days. Trying to start working on this today. :)

**Comment by miguelriemoliveira** [comment]
Hi André,

we're synced. I also did not have any time to devote to this until tonight
or tomorrow.
Miguel
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

is this output normal for the `agrob_view_optimization`? 

```
[ERROR] [1589210974.404574464]: PluginlibFactory: The plugin for class 'rviz_visual_tools/KeyTool' failed to load.  Error: According to the loaded plugin descriptions the class rviz_visual_tools/KeyTool with base class type rviz::Tool does not exist. Declared types are  rviz/FocusCamera rviz/Interact rviz/Measure rviz/MoveCamera rviz/PublishPoint rviz/Select rviz/SetGoal rviz/SetInitialPose rviz_plugin_tutorials/PlantFlag
[rospack] Error: no package given
[librospack]: error while executing command
[rospack] Error: no package given
[librospack]: error while executing command
[WARN] [1589210974.415142]: The 'use_gui' parameter was specified, which is deprecated.  We'll attempt to find and run the GUI, but if this fails you should install the 'joint_state_publisher_gui' package instead and run that.  This backwards compatibility option will be removed in Noetic.
[INFO] [1589210974.430104]: xacro_file is /home/andre/Source/catkin_ws/src/agrob/agrob_description/urdf/agrob.urdf.xacro
[INFO] [1589210974.433015]: prefix is initial_
[INFO] [1589210974.434224]: Reading xml xacro file ...
Unknown attribute "type" in /robot[@name='agrob']/transmission[@name='front_left_wheel_trans']
Unknown attribute "type" in /robot[@name='agrob']/transmission[@name='front_right_wheel_trans']
Unknown attribute "type" in /robot[@name='agrob']/transmission[@name='rear_left_wheel_trans']
Unknown attribute "type" in /robot[@name='agrob']/transmission[@name='rear_right_wheel_trans']
[INFO] [1589210974.451797]: Changing link names ...
[INFO] [1589210974.452871]: Changing joint names, parents and childs ...
[INFO] [1589210974.453733]: Writting new initial_robot_description parameter
[create_initial_robot_description-2] process has finished cleanly
log file: /home/andre/.ros/log/36517cbc-939c-11ea-921d-a0a4c5480909/create_initial_robot_description-2*.log
/opt/ros/melodic/lib/robot_state_publisher/robot_state_publisher: __name:=robot_state_publisher: command not found
[robot_state_publisher-3] process has died [pid 20426, exit code 127, cmd bash -c sleep 2; $4 $@ /opt/ros/melodic/lib/robot_state_publisher/robot_state_publisher __name:=robot_state_publisher __log:=/home/andre/.ros/log/36517cbc-939c-11ea-921d-a0a4c5480909/robot_state_publisher-3.log].
log file: /home/andre/.ros/log/36517cbc-939c-11ea-921d-a0a4c5480909/robot_state_publisher-3*.log
```

Also, I'm getting this error when launching the `test/sensor_pose_json_v2/main.py`:

```
Traceback (most recent call last):
  File "test/sensor_pose_json_v2/main.py", line 423, in <module>
    main()
  File "test/sensor_pose_json_v2/main.py", line 367, in main
    'x_scale': 'jac'})
  File "/home/andre/Source/OptimizationUtils/OptimizationUtils/OptimizationUtils.py", line 367, in startOptimization
    bounds=(bounds_min, bounds_max), method='trf', args=(), **optimization_options)
  File "/home/andre/.local/lib/python2.7/site-packages/scipy/optimize/_lsq/least_squares.py", line 871, in least_squares
    J0 = jac_wrapped(x0, f0)
  File "/home/andre/.local/lib/python2.7/site-packages/scipy/optimize/_lsq/least_squares.py", line 865, in jac_wrapped
    kwargs=kwargs, sparsity=jac_sparsity)
  File "/home/andre/.local/lib/python2.7/site-packages/scipy/optimize/_numdiff.py", line 400, in approx_derivative
    groups, method)
  File "/home/andre/.local/lib/python2.7/site-packages/scipy/optimize/_numdiff.py", line 500, in _sparse_difference
    df = fun(x) - f0
  File "/home/andre/.local/lib/python2.7/site-packages/scipy/optimize/_numdiff.py", line 348, in fun_wrapped
    f = np.atleast_1d(fun(x, *args, **kwargs))
  File "/home/andre/Source/OptimizationUtils/OptimizationUtils/OptimizationUtils.py", line 268, in internalObjectiveFunction
    self.vis_function_handle(self.data_models)  # call visualization function
  File "/home/andre/Source/OptimizationUtils/test/sensor_pose_json_v2/visualization.py", line 293, in visualizationFunction
    dataset_graphics = data['dataset_graphics']
KeyError: 'dataset_graphics'
```

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

view optimization is ok. 

The main.py has some problem. I will fix it as soon as possible (probably tonight).

**Comment by miguelriemoliveira** [comment]
@aaguiar96 ,

I cannot replicate your error. I launch like this:

`clear && test/sensor_pose_json_v2/main.py -json ~/datasets/agrob_first_test/data_collected.json -si -rv -ssf "lambda name: name in ['left_camera', 'vlp16']"`

How do you run the optimizer?

**Comment by aaguiar96** [comment]
> How do you run the optimizer?
```
roslaunch interactive_calibration agrob_view_optimization.launch
```

And the `main.py` exacly as you... 
Are you using [this](https://github.com/lardemua/AtlasCarCalibration/issues/141#issuecomment-623499169) dataset?

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

This morning I implemented the first version of the velodyne cost function.
I used pretty much the same approach that you used in the planar laser case, for the point-to-plane distance.

I will summarize the approach so that you can tell me if I understand what's going on:
- Compute the chessboard plane in the sensor reference frame
- Calculate the interception of the line that goes through the sensor origin to the point in the plane, with the estimated plane in the previous step
- Residual is the difference between the distance from the sensor origin to the intercepted point and the actual range measure taken by the sensor

I've got some questions now (sorry if they are stupid):
- The transformation between the chessboard and the sensor is computed locally on the `objective_function.py`. The "global/public" transformation are the ones present in the `chain` of the `json` file for each sensor. So, my question is, do we optimize the entire chain or the transformation between the target sensors?
- Also, all the objective functions are always computed in relation to the chessboard? Does this mean that the calibration procedure requires at least a camera?
- How do I test my objective function?

Also, I'm still stuck with that [bug](https://github.com/lardemua/AtlasCarCalibration/issues/149#issuecomment-626777827) ...

**Comment by miguelriemoliveira** [comment]
Hi André,

sorry for the late reply. I am not working today.

I just pushed a new version. Can you try now?

I will read your email more carefully later an dcall you ok?

Miguel
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

> Compute the chessboard plane in the sensor reference frame
> Calculate the interception of the line that goes through the sensor origin to the point in the plane, with the estimated plane in the previous step
> Residual is the difference between the distance from the sensor origin to the intercepted point and the actual range measure taken by the sensor
> I've got some questions now (sorry if they are stupid):

You are right. We call this the beam distance. It is an improvement over the orthogonal distance which was our first distance and is not used anymore. 

If you want to implment the orthogonal distance, this was how:

1. The goal is to bring the laser measurements to the chessboard reference frame. 
2. Thus we multiply the measurements (3D points in the lidar reference frame) by the root_T_sensor transformation, and later by the chessboard_T_root transformation.
3. With this we have transported those 3D points from the laser reference frame to the chessboard reference frame. 
4. Note that, since hte points are now on the chessboar reference frame, we just have to look at the z coordinate to get the distance to the chessboard plane.

> 
> The transformation between the chessboard and the sensor is computed locally on the objective_function.py. The "global/public" transformation are the ones present in the chain of the json file for each sensor. So, my question is, do we optimize the entire chain or the transformation between the target sensors?

Atomic transformations are the actual transforms that exist of the tf tree, as oposed to aggregate transformations.
No, we optimize only the atomic transformations which are marked for optimization. For example in your case, from your config.json:

```json
  "sensors": {
        "right_camera": {
            "link": "zed_right_camera_optical_frame",
            "parent_link": "zed_camera_center",
            "child_link" : "zed_right_camera_frame",
            "topic_name" : "/zed/zed_node/right/image"
        },

        "left_camera": {
            "link": "zed_left_camera_optical_frame",
            "parent_link": "zed_camera_center",
            "child_link" : "zed_left_camera_frame",
            "topic_name" : "/zed/zed_node/left/image"
        },

	"vlp16": {
	   "link": "velodyne",
	   "parent_link": "tower_link",
	   "child_link": "vlp16_frame",
	   "topic_name": "/velodyne_points"
	}
```
 you can see that we marked 3 atomic transformations to be optimized. All the others are untouched.

zed_camera_center to zed_right_camera_frame, 
zed_camera_center to zed_left_camera_frame
and
tower_link to  vlp16_frame"


> Also, all the objective functions are always computed in relation to the chessboard? 

Yes, we also estimate the pose of the chessboard (actually, lets start using the term pattern since it covers charucos and chessboards) during the optimization. As such, a first guess also for these parameters must be fed into the system to initialize the optimization. 

To solve this we use a [solvePNP](https://docs.opencv.org/2.4/modules/calib3d/doc/camera_calibration_and_3d_reconstruction.html#bool%20solvePnP(InputArray%20objectPoints,%20InputArray%20imagePoints,%20InputArray%20cameraMatrix,%20InputArray%20distCoeffs,%20OutputArray%20rvec,%20OutputArray%20tvec,%20bool%20useExtrinsicGuess,%20int%20flags)) function that needs at least one  camera to detect the pattern. 

Does this mean that the calibration procedure requires at least a camera?

Yes, for now yes. If / when we need we can come up with an alternative way of initializing the pattern poses.

> How do I test my objective function?
That's a difficult one, you can print the errors but that will only take you so far ... the best is to observe the system during optimization and to try to understand if the system is doing what we want it to do. Also you can view those graphs and check if the error is descending.
But this is one of the tricky parts, we actually are never sure ... @eupedrosa can you add something here?

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

> How do you run the optimizer?
> 
> roslaunch interactive_calibration agrob_view_optimization.launch
> And the main.py exacly as you...
> Are you using this dataset?

Yes, I am using the same dataset as you.

I run this test, it works ... 

https://youtu.be/fHILXskESA0

Perhaps your pycharm? Check the video ...

Also, about your question if the objective function is working. I told you it is hard to know when its working. Luckly, it is easier to know when it is not. Take a look and this image

![image](https://user-images.githubusercontent.com/11350998/81754015-ff7eb080-94ac-11ea-81b9-8c02e8e6993d.png)

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Thanks for the great replies. You clarified all my questions. :)

I'll try again to run the system using your video. This is weird...

Anyway, we are moving forward.
Do we talk tomorrow?

**Comment by aaguiar96** [comment]
Another thing @miguelriemoliveira 

I'm returning to my hometown this weekend. So, from monday onward I'm available to go to the Lab to record new rosbags.

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

we can zoom tomorrow 14h45?

Good news you can take a new bagfile. I was planning to give you the charuco we have but I am not sure I will go to Aveiro in the meantime.

When are you going to pass by Aveiro? Can you tell me a day / time?

Thanks,
Miguel

**Comment by aaguiar96** [comment]
> we can zoom tomorrow 14h45?

Yes @miguelriemoliveira, deal!

> When are you going to pass by Aveiro? Can you tell me a day / time?

Not sure yet the day and time. I let you know soon.

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira and @eupedrosa  

I was thinking... Isn't it enough to align the corners of the pattern with the four extrema beam points, instead of using all the extrema points?

**Comment by miguelriemoliveira** [comment]
Hi André.  The problem is the discretization in the point cloud . You can t
be sure that you  have a measurement of the real corner.

But if you want to start with that it is a good approximation

**Comment by aaguiar96** [comment]
Ok, I've implemented with the corners and obtained this result:

![optimization_1](https://user-images.githubusercontent.com/35901587/82582773-6d506980-9b8a-11ea-8475-63d971286cd6.png)
![optimization_2](https://user-images.githubusercontent.com/35901587/82582776-6e819680-9b8a-11ea-8f9b-4a618dd28791.png)

It's funny because the solution converges to an inclination on velodyne, but in reality the ZED contains an inclination of 11.3º.
@miguelriemoliveira, why does the ZED not moves? Is that because it is anchored?

**Comment by eupedrosa** [comment]
Do not trust in your eyes, you are calibrating the sensor constellation, not relative to the robot's body. 

> why does the ZED not moves? Is that because it is anchored?

YES

**Comment by miguelriemoliveira** [comment]
Looks good to me

**Comment by aaguiar96** [comment]
Hmmm... I checked the corner distances after converging, and some corners are still a bit distant.
Maybe some transformation is wrong. I'll check it out.

**Comment by eupedrosa** [comment]
How much is a "bit distant"?

**Comment by aaguiar96** [comment]
15cm... Clearly wrong, no?

**Comment by eupedrosa** [comment]
Yes.. Clearly wrong..

**Comment by aaguiar96** [comment]
Thanks, I'll check the math.

**Comment by aaguiar96** [comment]
Hi guys.

Have you tried to run the optimization with the new dataset?
I'm getting this error while executing the optimization:
```
clear && test/sensor_pose_json_v2/main.py -json /home/andre/Documents/rosbags/agrob_second_dataset/data_collected.json -si -rv -ssf "lambda name: name in ['left_camera', 'right_camera', 'vlp16']"
```
```
Argument list={'skip_vertices': 1, 'use_incomplete_collections': False, 'single_pattern': False, 'json_file': '/home/andre/Documents/rosbags/agrob_second_dataset/data_collected.json', 'collection_selection_function': None, 'view_projected_vertices': False, 'ros_visualization': True, 'z_inconsistency_threshold': 0.05, 'sensor_selection_function': <function <lambda> at 0x7f145f8e2f50>, 'show_images': True, 'view_optimization': False}


Collections studied:
 [u'11', u'10', u'13', u'12', u'15', u'14', u'17', u'16', u'18', u'1', u'0', u'3', u'2', u'5', u'4', u'7', u'6', u'9', u'8']
Traceback (most recent call last):
  File "test/sensor_pose_json_v2/main.py", line 426, in <module>
    main()
  File "test/sensor_pose_json_v2/main.py", line 148, in main
    dataset_sensors['chessboards'], dataset_chessboard_points = createChessBoardData(args, dataset_sensors)
  File "/home/andre/Source/catkin_ws/src/OptimizationUtils/test/sensor_pose_json_v2/chessboard.py", line 278, in createChessBoardData
    ret, rvecs, tvecs = cv2.solvePnP(objp, corners, K, D)
cv2.error: OpenCV(4.2.0) /io/opencv/modules/calib3d/src/solvepnp.cpp:754: error: (-215:Assertion failed) ( (npoints >= 4) || (npoints == 3 && flags == SOLVEPNP_ITERATIVE && useExtrinsicGuess) ) && npoints == std::max(ipoints.checkVector(2, CV_32F), ipoints.checkVector(2, CV_64F)) in function 'solvePnPGeneric'
```

I checked and the dataset has always more than 30/40 pattern points...
Any idea?

**Comment by aaguiar96** [comment]
Also,

Because of the bug I could not do many debug, but what I referenced here:
> Hmmm... I checked the corner distances after converging, and some corners are still a bit distant.
Maybe some transformation is wrong. I'll check it out.

might be related with wrong chessboard dimensions in the config file for the first dataset.
Also (and this is just a doubt), since in the config file we always specify the number of squares **minus 1**, shouldn't we add **1** while computing the limit points here (for the cases where the pattern is a charuco):

https://github.com/miguelriemoliveira/OptimizationUtils/blob/a2055032709a80556670ae989228cdedb7cef87f/test/sensor_pose_json_v2/chessboard.py#L37

**Comment by aaguiar96** [comment]
Another thing (sorry about the spam...)

Since the ZED camera has an inclination in reality which is not represented in the agrob description, shouldn't it be better to anchor the velodyne?

**Comment by eupedrosa** [comment]
About the error, try replacing
```python
cv2.solvePnP(objp, corners, K, D)
```
with
```python
cv2.solvePnP(objp, np.array(corners, dtype=np.float32), K, D)
```

The error that you have is usually a simptom of bad numpy type in one of the arguments.

> might be related to wrong chessboard dimensions in the config file for the first dataset.

Unrelated.

> Also (and this is just a doubt), since in the config file we always specify the number of squares minus 1, shouldn't we add 1 while computing the limit points here? (for the cases where the pattern is a charuco):

What limits are you taking about?

> Since the ZED camera has an inclination in reality which is not represented in the agrob description, shouldn't it be better to anchor the velodyne?

It is neither better nor worse. But yeah, you can change to the velodyne.

**Comment by aaguiar96** [comment]
Thanks @eupedrosa 

> Unrelated.

Why do you say so?
I am using the pattern limit points that were already calcutated on the optimization code.
However, these calculations use the pattern sizes set on the config, that were wrong in the first dataset since I did not measure the actual pattern square sizes. 

> What limits are you taking about?

The pattern limits calculated on the optimization framework. In the link that I sent before.

> It is neither better nor worse. But yeah, you can change to the velodyne.

Visually, the ZED will never become inclined if it is anchored right? Of course if the optimization is right, the solution is equal for both cases! :)

**Comment by eupedrosa** [comment]
>>Unrelated.
>
>Why do you say so?

Because the number of `objp` should be equal to the number of corner. Even if the number was wrong for the given pattern they should match. 

However, using a charuco they may not match. Charuco has partial detections. That can also be the proble. I solved that in `general_hand_eye.py`

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 and @eupedrosa ,

sorry, I was with the kids until now. From what I read there is a path to a solution already.

Do you need something from my side?

**Comment by aaguiar96** [comment]
> About the error, try replacing
> 
> ```python
> cv2.solvePnP(objp, corners, K, D)
> ```
> 
> with
> 
> ```python
> cv2.solvePnP(objp, np.array(corners, dtype=np.float32), K, D)
> ```

@miguelriemoliveira  and @eupedrosa, Eurico's suggestion did not your.
I still got the same error when running the optimization in the new dataset.
Can you try it out and check if you got it as well? :)

**Comment by aaguiar96** [comment]
Also, 

I have the same question that I had before...
I'm using the pattern extrema points calculated in `chessboard.py` that are also used in the `LaserScan`.

I was checking why the corners of the pattern and the corners of the points of the lidar on the patter are so distant and I noticed that:

- The euclidean distance between the corners given by the extrema points of the pattern and the extrema points of the lidar points on the pattern have the same width and height (which is good) 
- But, the respective corners do not match. For example, the corners of the pattern given by PnP are [-0.29, 0.2], [0.29, 0.2] and the corners of the lidar points on the pattern are [-0.16, -0.10],[0.43,0.27].

Of course I may be doing something wrong, but my question is. @miguelriemoliveira, are the pattern limit points calculation general enough to be used for both a chessboard and a charuco?
For example, did you saw what I mentioned here?

> Also (and this is just a doubt), since in the config file we always specify the number of squares minus 1, shouldn't we add 1 while computing the limit points here? (for the cases where the pattern is a charuco):
https://github.com/miguelriemoliveira/OptimizationUtils/blob/a2055032709a80556670ae989228cdedb7cef87f/test/sensor_pose_json_v2/chessboard.py#L37

Thanks guys!

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

I will be teaching tomorrow and wednesday morning. @eupedrosa can you try to meet with  @aaguiar96 to help him with this? 

If you can't solve it until wednesday afternoon, then we can meet and I will try to help ok?

**Comment by eupedrosa** [comment]
I think I can help. @aaguiar96 tomorrow after 17h I am available. You you still need my help just send me an email to set up a meeting.

**Comment by aaguiar96** [comment]
Nice @eupedrosa 

Can you meet at 18h00? We can also talk about the colada file generation! :)

**Comment by eupedrosa** [comment]
Ok, 18h00.

**Comment by aaguiar96** [comment]
#159 has to be solved first.

**Comment by aaguiar96** [comment]
With the new dataset. Not looking good by now.

![bad_res](https://user-images.githubusercontent.com/35901587/83009311-7a4bdd80-a00e-11ea-8155-41bad8d78585.png)

**Comment by miguelriemoliveira** [comment]
You already implemented #159?

**Comment by aaguiar96** [comment]
> You already implemented #159? These results are for that? We can talk at 15h?

I did not @miguelriemoliveira. These results are still with the problematic approach.

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Using the code that you had for visualizing the limit points, I got this:

![calib](https://user-images.githubusercontent.com/35901587/83051653-370d6100-a046-11ea-8a9b-fb0cf63b2b83.png)

I have to transform `dataset['chessboards']['limit_points']` to the root frame right?

**Comment by miguelriemoliveira** [comment]
No, they should be on the local frame. Rviz transforms the points using the tfs...

**Comment by aaguiar96** [comment]
Ok, got them. 

![points](https://user-images.githubusercontent.com/35901587/83054412-48f10300-a04a-11ea-974c-372af2e11130.png)

I'll start to work on the LiDAR corners calculation! :)

**Comment by miguelriemoliveira** [comment]
Great. You should also draw the horizontal lines ...

**Comment by aaguiar96** [comment]
You commented the horizontal lines on the calculation of the `limit_points`. Can I uncomment?

**Comment by eupedrosa** [comment]
What are the news????? What did I miss?

**Comment by miguelriemoliveira** [comment]
Sure

**Comment by aaguiar96** [comment]
> What are the news????? What did I miss?

We noticed that the corners are not being well calculated.
Also, we are now drawing the corners of the Lidar points and the pattern limits.

I'll start the fix of the corners calculation.

**Comment by eupedrosa** [comment]
Did you discoverd why it was being miscalculated?

**Comment by miguelriemoliveira** [comment]
Hi @eupedrosa ,

several things: the fact that we were using a transformation from the lidar to the chessboard and the technique of finding the max in x and y also does not work unless the pattern is aligned with those axes, which would be a misleading assumption.

**Comment by aaguiar96** [comment]
Hi guys, I'll just report here the latest attempt to compute the limit points of the LiDAR.
It is still not perfect.
The idea was to: Compute local minimum/maximum - Save - Delete to do the same in the next iteration.

![rviz_screenshot_2020_05_28-18_28_05](https://user-images.githubusercontent.com/35901587/83173910-8076b280-a111-11ea-8320-7997fa01471a.png)


This approach still has problems.
I'll try to implement sort of a floodfill. I think this will work better.

**Comment by miguelriemoliveira** [comment]
Hi André,

How about a minimum area rectangle fitting? Opencv has a function for this.

The problem is you would have to run this on a 2d projection ...

If you want we can discuss about it

**Comment by aaguiar96** [comment]
> How about a minimum area rectangle fitting? Opencv has a function for this.

That's a good idea, I can try to work on it.

However, I came up with a solution that, despite being a bit archaic, works pretty well.
So, my assumption was that the closest point of the LiDAR to an arbitrary point outside the pattern will be an extrema point.
So, what I did was to compute a circle of points outside the pattern (as many as I want, since I  can configure the step), and compute the closest point in the LiDAR cloud to each one of them.

It's worth noting that this solution does not ensure that we find **all** the extrema points, but almost all of them if the `step` is small enough.

I'll show here the result for three different collections.
The circle is just represented to show the source points (I'll delete them from the visualization).

![rviz_screenshot_2020_05_29-14_56_44](https://user-images.githubusercontent.com/35901587/83268250-3d721900-a1bd-11ea-8cfb-f3845e5c268b.png)
![rviz_screenshot_2020_05_29-14_56_20](https://user-images.githubusercontent.com/35901587/83268252-3e0aaf80-a1bd-11ea-98b6-74e97a8d1004.png)
![rviz_screenshot_2020_05_29-14_55_40](https://user-images.githubusercontent.com/35901587/83268253-3ea34600-a1bd-11ea-8093-0b1703c6b4b5.png)


What do you think? I think @eupedrosa will not like it since it does not use linear algebra eheheh :)

**Comment by miguelriemoliveira** [comment]
Hi André,

Its a great idea ... ans i think it will find all points if your resolution
is high enough.

I thought you wanted to get the corners, not all the border points. Are you
going to use all border points?

Ans dont worry about Eurico. He's been crancky because he s writting the
paper : - )

**Comment by aaguiar96** [comment]
Thanks @miguelriemoliveira 



> I thought you wanted to get the corners, not all the border points. Are you going to use all border points?

I turns out that I had more trouble projecting a solution for finding the corners than the `limit points`, and since the higher the number of points, the better, why not?
Don't you agree? Is ok to use all the extrema points right?

**Comment by eupedrosa** [comment]
I have to say,  I do not understand you aproach @aaguiar96 .. Maybe I am to cranky.. Can you make a diagram....

**Comment by aaguiar96** [comment]
Yes (sorry about the bad diagram...)

The idea is to compute a discrete circle and find the closest point in the pattern to each point in the circle - it will always be an extrema point.

Like this:

![diag](https://user-images.githubusercontent.com/35901587/83276065-fb9aa000-a1c7-11ea-9d62-7f268e3a1a66.png)

**Comment by eupedrosa** [comment]
https://en.wikipedia.org/wiki/Alpha_shape

The obvious existing python package https://pypi.org/project/alphashape/

**Comment by aaguiar96** [comment]
Good suggestion. Do you think I should change to that approach?

This one is working, but I can try that too.

**Comment by miguelriemoliveira** [comment]
Crancky ... definetly  crancky : - )

**Comment by eupedrosa** [comment]
> Good suggestion. Do you think I should change to that approach?
This one is working, but I can try that too.

I think you should try. 
Your method may work but I have several issues with it. Mainly its performance...

**Comment by aaguiar96** [comment]
>  I think you should try.

Sure.


>  Your method may work but I have several issues with it. Mainly its performance...

I think that's not a problem. I only execute this method once for each collection (the extrema points are only computed once at the beggining of the optimization), and I tried with a step of 0.05 degrees (really low) and it only took about ~2/3 seconds for all the 19 collections.

**Comment by eupedrosa** [comment]
Try alpha shape and then report the times for comparison.

**Comment by miguelriemoliveira** [comment]
Hi,

My opinion is we should focus on the main goal. Performance, especially for
something which runs only once is not important.

André,  if you can try Eurico's suggestion but my suggestion is dont loose
too much time with it.


Regards
Miguel

**Comment by aaguiar96** [comment]
Ok, thanks!

**Comment by aaguiar96** [comment]
I @miguelriemoliveira 

Can you check if this is what you had in mind?
The projection of the pattern plane are the points in the darker blue.

![proj](https://user-images.githubusercontent.com/35901587/83567299-4ba39a80-a519-11ea-9001-3b138da9848a.png)

**Comment by miguelriemoliveira** [comment]
Hi André,

I do not understand the drawing very well. But I think that is not what I
had in mind.

Perhaps we could talk. I have classes / meeting all day. How about tonight?
If you cannot, then tomorrow.

Miguel
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by aaguiar96** [comment]
Yes @miguelriemoliveira 

I can talk today around 19h00, or tomorrow anytime but 11h30.

**Comment by miguelriemoliveira** [comment]
Hi André,

not sure if I can.

I will call you if I can.

*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Do you have any utility to built a `tf::transform` from a translation vector and a rotation matrix?

**Comment by miguelriemoliveira** [comment]
Hi,

check here if it has what you need

https://github.com/miguelriemoliveira/OptimizationUtils/blob/master/OptimizationUtils/utilities.py

Miguel
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by aaguiar96** [comment]
Namely, this (@eupedrosa using your idea)

![Figure_1](https://user-images.githubusercontent.com/35901587/83775078-73f8d980-a67e-11ea-91d3-5f177fa7d3bd.png)

**Comment by eupedrosa** [comment]
> Do you have any utility to built a tf::transform from a translation vector and a rotation matrix?

Are you using ROS inside the objective function?

**Comment by eupedrosa** [comment]
> Namely, this (@eupedrosa using your idea)

Is it slow? fast?

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

We can talk in 10 or 15 mins. I will call you ok?

**Comment by aaguiar96** [comment]
> We can talk in 10 or 15 mins. I will call you ok?

Ok! :)

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

I tried the inverse transformation but it does not get better...
I push to `agrob` to set up the `calibration.launch` and to `atom` the most recent changes.

With the `calibrate` script I am getting and error in the visualization.
```
Traceback (most recent call last):
  File "/home/andre/Source/catkin_ws/src/atom/atom_core/scripts/calibrate", line 549, in <module>
    main()
  File "/home/andre/Source/catkin_ws/src/atom/atom_core/scripts/calibrate", line 477, in main
    graphics = visualization.setupVisualization(dataset_sensors, args)
  File "/home/andre/Source/catkin_ws/src/atom/atom_core/src/atom_core/visualization.py", line 60, in setupVisualization
    description_file,_,_ = uriReader(dataset['calibration_config']['description_file'])
KeyError: 'description_file'
```

If you got news let me know.
I left some questions for you in the `calibrate` script (just search for "@migueloliveira")

Thanks in advance! :)

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

Q&A:

> import OptimizationUtils.utilities as optimization_utilities  # @migueloliveira should we use this here?

For now this is confusing because we have an utilities in optimization_utils (which will continue to exists and be used) and in atom_calibration. I will fix it later, for now leave it as is.

>  - Compute polygonal boundary using convex hull (@migueloliveira: we should add these to requirements.txt right?)


yes

>  - @migueloliveira the code bellow is the approach to find the extrema points with the circle
>    but it is still using raw velodyne points. So, don't look at it now :)

My eyes are shut.

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 again,

> Traceback (most recent call last):
>   File "/home/andre/Source/catkin_ws/src/atom/atom_core/scripts/calibrate", line 549, in <module>
>     main()
>   File "/home/andre/Source/catkin_ws/src/atom/atom_core/scripts/calibrate", line 477, in main
>     graphics = visualization.setupVisualization(dataset_sensors, args)
>   File "/home/andre/Source/catkin_ws/src/atom/atom_core/src/atom_core/visualization.py", line 60, in setupVisualization
>     description_file,_,_ = uriReader(dataset['calibration_config']['description_file'])
> KeyError: 'description_file'

this is because your dataset is deprecated. We have added new stuff to the config.yml and your dataset was created before that.

As a one time patch you can add the parameter description_file in the data_collected,json of the dataset you are using (it should be inside "calibration_config", since this is where the copy of config.yml is)

**Comment by miguelriemoliveira** [comment]
... now I will try to see what's going wrong ...

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

I am having this error:

> Traceback (most recent call last):
>   File "/home/mike/catkin_ws/src/calibration/atom/atom_core/scripts/calibrate", line 554, in <module>
>     main()
>   File "/home/mike/catkin_ws/src/calibration/atom/atom_core/scripts/calibrate", line 240, in main
>     c, normal = fitPlaneLTSQ(points[:, 0:3])
>   File "/home/mike/catkin_ws/src/calibration/atom/atom_core/scripts/calibrate", line 68, in fitPlaneLTSQ
>     (a, b, c), resid, rank, s = np.linalg.lstsq(G, Z, rcond=None)
>   File "/usr/lib/python2.7/dist-packages/numpy/linalg/linalg.py", line 1953, in lstsq
>     0, work, -1, iwork, 0)
> TypeError: a float is required

too tired right now to figure out whats wrong ... but its strange, you were using this function without a problem ,,,

**Comment by miguelriemoliveira** [comment]
Ok, found the problem, calling the funciton with the rcond argument:

`(a, b, c), resid, rank, s = np.linalg.lstsq(G, Z, rcond=None)`

passed to 

`(a, b, c), resid, rank, s = np.linalg.lstsq(G, Z)`

and now it works

**Comment by miguelriemoliveira** [comment]
Hi again,

I have a problem because of updating the husky software pakcage 

> Parsing description file /home/mike/catkin_ws/src/calibration/agrob/agrob_description/urdf/agrob.urdf.xacro
> Executing command: xacro /home/mike/catkin_ws/src/calibration/agrob/agrob_description/urdf/agrob.urdf.xacro -o /tmp/description.urdf
> error: 'bool' object has no attribute 'lower' 
> when evaluating expression 'topbar_needed_kinect.lower() == 'true' or topbar_needed_kinect == '1' or topbar_needed_realsense == 'true' or topbar_needed_realsense == '1''
> when processing file: /home/mike/catkin_ws/src/calibration/husky/husky_description/urdf/husky.urdf.xacro

solved it with removing the **topbar_needed_kinect.lower() == 'true' or** part, i.e.:

```
<!--  <xacro:if value="${topbar_needed_kinect.lower() == 'true' or topbar_needed_kinect == '1' or topbar_needed_realsense == 'true' or topbar_needed_realsense == '1'}">-->
  <xacro:if value="${topbar_needed_kinect == '1' or topbar_needed_realsense == 'true' or topbar_nee
```

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Thanks for the fixes. 
I'll look at this again in the afternoon.
If you have some news let me know. :)

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

I'll start to work on this now.
Call me if you want to join. :)

**Comment by miguelriemoliveira** [comment]
Hi André

I can't right now. I will work only tonight.

I will call you to see whats new.

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

The `test_plane_fiting.py` is working...

![Figure_1](https://user-images.githubusercontent.com/35901587/83906961-6ca7fd80-a75c-11ea-903b-52969486c67f.png)

**Comment by miguelriemoliveira** [comment]
Hum, if it works here, it must work on the other side. Can you plot the z axis here ... and then on the other side?

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

I think I solved the problem with the z axis. The problem was defining the X axis. We were using two points from the cloud, which do not necessarely lie on the plane. I added a section with a code that works. Please take a look and fuse with your code (which I commented) and push a clean version. 

![image](https://user-images.githubusercontent.com/11350998/83931575-5b2f1780-a795-11ea-8750-3d345b6a6010.png)

Now you have a correct reference system. Move on to the getting the limit points using the alpha shape. 

If I have time this weekend I will look into the drawing of the pattern (from the config.yml data) part.

**Comment by aaguiar96** [comment]
Oh... Ok I got it. The trick was to convert the two points into the plane using the plane equation right?
Ok, always learning. Thanks, @miguelriemoliveira :)

I'll handle the rest of it as soon as possible.

**Comment by miguelriemoliveira** [comment]
Right, I forgot to mention that. because if we don't we are not sure that the x axis lies in the plane. Tell me if you need more help.

**Comment by aaguiar96** [comment]
Right, that makes sense! :)

Now, I have the local coordinate system in reference to the velodyne.
So, if I apply the inverse transformation, I'll have a plane aligned with the velodyne axis, then I can project the 3D points to 2D ignoring the `z` axis. No, it's safe to project to 2D since we have a local coordinate system, right?
Just to make sure I'm following your reasoning.

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,


> Now, I have the local coordinate system in reference to the velodyne.

yes, the local pattern reference system w.r.t the velodyne reference system.

> So, if I apply the inverse transformation, I'll have a plane aligned with the velodyne axis, then I can project the 3D points to 2D ignoring the z axis. No, it's safe to project to 2D since we have a local coordinate system, right?

Yes, you must bring the points from the velodyne reference system (where they are when you get the data) to the local pattern reference system by multiplying by the transformation (I never now if its by the normal or the inverse)

> Just to make sure I'm following your reasoning.

You are.

> The result of the inverse transformation - points aligned with velodyne.

The drawing I do not understand because, as I told you, we can only draw points in a single reference system. You have points drawn in difference reference systems in this drawing, no?

**Comment by aaguiar96** [comment]
> The drawing I do not understand because, as I told you, we can only draw points in a single reference system. You have points drawn in difference reference systems in this drawing, no?

The two set of points are drawn in velodyne frame in `rviz`. Since I applied the inverse transformation, the darker blue points are aligned with velodyne. Makes sense, right?

**Comment by miguelriemoliveira** [comment]
Hum ... I don't think it makes sense to draw at all. We can talk by phone. Call me when you want.

**Comment by aaguiar96** [comment]
> Hum ... I don't think it makes sense to draw at all. We can talk by phone. Call me when you want.

I'll call you soon! :)

So, using `alphashape` I don't see a straightforward solution to extract to limit points. It computes a low number of vertices, so some processing to the polygon is required to isolate the limit points.
I used my initial approach of the circle fitting, and it works pretty well. 
I think I'll use this approach now, and send the `alphashape` to future work.

Here's some results for different collections:

![border1](https://user-images.githubusercontent.com/35901587/83973367-e1ef0c00-a8dd-11ea-8678-d902e68bf873.png)
![border2](https://user-images.githubusercontent.com/35901587/83973370-e3203900-a8dd-11ea-9596-6d7777e2ab7a.png)
![border3](https://user-images.githubusercontent.com/35901587/83973371-e4516600-a8dd-11ea-9f3b-1d85b7616e9e.png)

I think we're ready to had this feature to the objective function! :)

**Comment by aaguiar96** [comment]
Using dynamic circle radius computed with the maximum distance between a pair of points in the 2D vector:

![Figure_1](https://user-images.githubusercontent.com/35901587/84067362-3cac6480-a9bf-11ea-97da-d9935037ed07.png)
![dynamic](https://user-images.githubusercontent.com/35901587/84067367-3e762800-a9bf-11ea-92e0-d56e2012e7eb.png)

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

is there a way to parameterize the angular step?

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Do you mean set it as an input parameter or computed it in a automatic manner as I did for the radius?

**Comment by miguelriemoliveira** [comment]
I meant as an input parameter. I don't know how you would do it automatically.

In any case, I think this is a very low priority, since this is done offline once, you can just hardcode a high enough angular step to always get all measurements.

**Comment by aaguiar96** [comment]
> yes, right. I would say:
> 
> ```
> for each point in limit_points_detected_in_the_cloud:
>     compute minimum distance to all the sampled points in the pattern frame
>     use the minimum as the error for that point
> ```

That was what I had in mind too! :)

> 
> with all the changes that I was making in the patterns I may have introduced some bug ,,, let me know if the optimizer is doing awkward stuff.

Ok @miguelriemoliveira 
Is the rviz configuration automatically showing the patterns on your side? I had to manually add them.

**Comment by miguelriemoliveira** [comment]
yes it is all automatic. After git pull you will have to rerun the configure of your package.

**Comment by aaguiar96** [comment]
> After git pull you will have to rerun the configure of your package.

That I did not know. Thanks!

**Comment by miguelriemoliveira** [comment]
Well, it is not always. But I have recently been making additions to the configure script ...

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

In which reference frame are the limit points that you computed for the pattern?
Both set of limit points have to be in the same referential to compute the difference between them right?

**Comment by miguelriemoliveira** [comment]
Exactly. The reference frame of the points is the local pattern reference frame. That is why you have no z coordinate for the points, since its all zero. You can create a 3d np array like this:

https://github.com/lardemua/atom/blob/3f78cbaa0e927ccc185b37c76b80dbd1ea14218e/atom_core/src/atom_core/objective_function.py#L256-L262

I would say the best way is to convert the lidar points (detected limit points in the lidar data) to the pattern local frame. It is easier for example to compute orthogonal error (just look at the z coordinate of those points).

Another less interesting alternative is to bring the pattern points to the lidar reference frame.

The least interesting is to bring both to a different reference frame. Just adds additional computations without any real gain.

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

problem solved in the camera. Now you can try to optimize a camera-velodyne tandem. Can't wait for the first results ...

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Not working by now... Maybe we need to visualize what's happening as you said.
Also, I'm trying with a single collection? Is that enough?

![rviz_screenshot_2020_06_11-00_37_17](https://user-images.githubusercontent.com/35901587/84329185-ef292680-ab7b-11ea-9833-8ce9f76abafe.png)

**Comment by miguelriemoliveira** [comment]
Hi,

about the visualization, I suggest that you first try it out without adding
more visualization tools. Only if it fails and we need to debug we should
consider more visualization.

One collection is a good way to have a feeling about if the optimization is
doing what it should be doing in a quick way. Try first with one, then if
it works use more.


*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by aaguiar96** [comment]
Ok @miguelriemoliveira 

It is not working with a single collection and the approach that we talked.
I'll try to find if there is any bug.

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

I will try to take a look. Improved the configure_calibration_pkg. You should run reconfigure.

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

I tried to implement the longitudinal distance but have to stop before I finish. Take a look at this part in the objective function, thats where the juice is. 

>  ------------------------------------------------------------------------------------------------
>  --- Pattern Extrema Residuals: Distance from the extremas of the pattern to the corners of the cloud
>  ------------------------------------------------------------------------------------------------

Also, you can search in the atom_core for Andre and you'll see several comments I left for you.

I think I advanced a bit, lets see if you cen get some results tomorrow.

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Thanks. Unfortunately, I forgot to commit my changes since I did not have good results.
Your code looks very similar to what I had, but let's see if it works better! :)

**Comment by aaguiar96** [comment]
It's given the same result.
I committed the code now so that you can look at it if you want.

I'll do some debug now!

**Comment by miguelriemoliveira** [comment]
Hum ... so it does not work.

One trick that helped me was to consider a single detected point. In you
detections just take the first in the list.

Then only one limit point is drawn, amd only that one should, over the
optimization, come closer to one of the ground truth limit points.

Also, use only two sensors and a single collection.

The general idea is to rrduce the problem as much as possible until it is
easier to find the problem.

**Comment by miguelriemoliveira** [comment]
Today I am not working until night.

I will check email...

**Comment by miguelriemoliveira** [comment]
Hi, I was investigating the problem with the agrob calibration and came to the conclusion that the calibration of two agrob cameras does not work. Not even a calibration with a single camera works:

`roslaunch agrob_calibration calibrate.launch dataset_file:=~/datasets/agrob/calibration/data_collected.json csf:='lambda x: int(x)<1' ssf:='lambda name: name in ["right_camera"]'`

If the camera part is not working, the lidar never will, since we always use a camera with the lidar.

After doing plenty of debug I found this:

![image](https://user-images.githubusercontent.com/11350998/84452136-dccaed80-ac4c-11ea-9770-ec9e00a5fe69.png)

On top is a part of the labels of a chessboard, and below the labels of a charuco.
The first has x growing first, while the second has y growing first.

Not sure if this is a problem. @eupedrosa , how did you generate the ground truth points in the hand eye?

**Comment by miguelriemoliveira** [comment]
I was thinking about this and I think the reason why the x and y and changed in because the pattern in the agrob bagfile is in a portrait orientation, while the other bagfile has the pattern in a landscape orientation. So perhaps the bug is somewhere else ...

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira 

Ok, at least we know what is the problem.
Do you want to talk today about this?

**Comment by miguelriemoliveira** [comment]
Hi André, good morning.

If you want we can talk now before the kids wake up.



*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by miguelriemoliveira** [comment]
Another important point: I tested the optimization of two cameras using the atlascar2 dataset and it is working very  well. So what puzzles me is that, if its the image part of the objective function is working for the atlascar2, then we should not touch it ... I am puzzled

**Comment by aaguiar96** [comment]
Ok, 5 minutes and I'm in zoom.



> Another important point: I tested the optimization of two cameras using the atlascar2 dataset and it is working very well. So what puzzles me is that, if its the image part of the objective function is working for the atlascar2, then we should not touch it ... I am puzzled

What changed from atlascar to agrob was the pattern type right? Does the eye-in-hand case use the same objective function for the cameras?

**Comment by eupedrosa** [comment]
> What changed from atlascar to agrob was the pattern type right? Does the eye-in-hand case use the same objective function for the cameras?

I think it is not the same. At the time wrote it from scratch to better understand it. But it the theory is the same.

**Comment by aaguiar96** [comment]
> I think it is not the same. At the time wrote it from scratch to better understand it. But it the theory is the same.

Ok, since the cost function works for a chessboard, in theory, it should work for a charuco right?

**Comment by eupedrosa** [comment]
Yes, it should not matter if it is chessboard or charuco.

**Comment by aaguiar96** [comment]
> Yes, it should not matter if it is chessboard or charuco.

Ok thanks, I'll make some debug and look at your cost function too!

**Comment by aaguiar96** [comment]
Hi guys

Just noticed that the ZED camera does not appear in the agrob mesh for some reason...
@miguelriemoliveira have you noticed that too?

**Comment by miguelriemoliveira** [comment]
I did not notice. Please create an issue for that and I will fix it.

**Comment by aaguiar96** [comment]
Hi @eupedrosa and @miguelriemoliveira 

The chain of transforms in each collection for each sensor integrates all the transformations from the `base_link` into the `sensor` right?

If so, here, why do you use the inverse transformation (`sensor` to `base_link`) to convert the points in the `base_link` frame to the `sensor` frame? Shouldn't it be the opposite?... Since it's working for atlascar, 99% sure my reasoning is not correct. :)

https://github.com/lardemua/atom/blob/62b0bad18264c397bfcf5cf089d43200bee67efd/atom_core/src/atom_core/objective_function.py#L147-L149

**Comment by miguelriemoliveira** [comment]
Hi @aaguiar96 ,

I did a small matlab example. Can you take a look and tell me if you think that we should use the inverse of the direct transform?

Then I will tell you my conclusion.

**Comment by miguelriemoliveira** [comment]
its here

https://github.com/lardemua/atom/blob/master/atom_calibration/scripts/DirectionOfTransforms.m

**Comment by aaguiar96** [comment]
Thanks @miguelriemoliveira 

I'll take a look!

**Comment by miguelriemoliveira** [comment]
Ok, I found out the problem. It had to do with the fact that we were using only four  corners  and, for some reason, we were not associating those four points with the ground truth points. 

I used all the points in the pattern and now the optimization works.

NOTE: I find that collection 3 does not have correct detections, both for the left and right cameras.
Look at the squares, those are the ground truth (the detections)

![image](https://user-images.githubusercontent.com/11350998/84507651-15081580-acb9-11ea-9bbb-84af29f5b72d.png)

I am not sure if this is wrong or just a partial detection (@eupedrosa , can you help?)

The image objective function I am pushing works for complete detections, not for partial detections.

At least for now, if you remove collection 3, you can test out the lidar.

**Comment by aaguiar96** [comment]
Ok @miguelriemoliveira 

Great, thanks! I will test then, for a single collection first.

**Comment by eupedrosa** [comment]
> I am not sure if this is wrong or just a partial detection (@eupedrosa , can you help?)

It may not be wrong if what you are showing is the uncalibrated camera.
Partial detections are working in hand-eye.

**Comment by aaguiar96** [comment]
Hi @miguelriemoliveira and @eupedrosa 

I think it's working... :)

![rviz_screenshot_2020_06_12-15_08_18](https://user-images.githubusercontent.com/35901587/84511580-a4fc8e00-acbe-11ea-8bbc-28dc41c2b770.png)

The only think that is not making sense is the velodyne being so high. But definitly the cloud points and pattern points are aligned!

**Comment by eupedrosa** [comment]
> The only think that is not making sense is the velodyne being so high.

The sensor is anchored. 
And you have to ignore the robot visuals, they are not part of the optimization. You are optimizing the sensors topology.

**Comment by eupedrosa** [comment]
By the way, nice work :+1: 
@miguelriemoliveira should now be free to finish some writing ...

**Comment by aaguiar96** [comment]
> The sensor is anchored.
> And you have to ignore the robot visuals, they are not part of the optimization. You are optimizing the sensors topology.

Ok. It would be nice to have a way to see the calibrated system in the correct spot.
For example, the ZED camera has an inclination of ~11 degrees, shouldn't we be able to notice that looking for the calibration?

> By the way, nice work +1

Thanks, here's another result:

![rviz_screenshot_2020_06_12-15_13_58](https://user-images.githubusercontent.com/35901587/84512454-ee99a880-acbf-11ea-93d1-0c766696fbcf.png)

**Comment by eupedrosa** [comment]
Set the velodyne as the anchored sensor....

And, where is the ZED camera?

**Comment by aaguiar96** [comment]
> And, where is the ZED camera?

That's what I'm trying to understand. Disappeared from the description for some reason...

**Comment by miguelriemoliveira** [comment]
Hi all,

it seems we are advancing. Great. @eupedrosa , I promise tonight I will look into the paper.

The velodyne is not on the correct position, right? I don't mean w.r.t the robot mesh (since as @eupedrosa said we cannot use this), but w.r.t to the camera's reference frame. Is the distance from the velodyne to the camera's ref frame realistic?

**Comment by aaguiar96** [comment]
> The velodyne is not on the correct position, right? I don't mean w.r.t the robot mesh (since as @eupedrosa said we cannot use this), but w.r.t to the camera's reference frame. Is the distance from the velodyne to the camera's ref frame realistic?

I don't think so... It would help if we see the camera. @miguelriemoliveira maybe the parsing of the xacro file is with some problem no?

**Comment by eupedrosa** [comment]
By the way @aaguiar96, does the ZED camera has a factory intrinsic parameters?? Did you check if `camera_info` has realistic values?

**Comment by miguelriemoliveira** [comment]
I will try to look into that issue.
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
office: 22.2.18                 |    ext: 23887
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by aaguiar96** [comment]
> By the way @aaguiar96, does the ZED camera has a factory intrinsic parameters?? Did you check if `camera_info` has realistic values?

They are realistic, but I don't if they are correct.
Here's the information about that.

https://support.stereolabs.com/hc/en-us/articles/360007497173-What-is-the-calibration-file-

They say the intrinsics are generated during the manufacturing process...

**Comment by eupedrosa** [comment]
It looks like it has: [What is the calibration file?](https://support.stereolabs.com/hc/en-us/articles/360007497173-What-is-the-calibration-file-)

But we need to know if this is done in ROS

**Comment by aaguiar96** [comment]
> But we need to know if this is done in ROS

I think so, look at [this](https://github.com/stereolabs/zed-ros-wrapper/issues/39#issuecomment-227371066).

**Comment by eupedrosa** [comment]
Can you put here the output of a `camera_info`?

**Comment by aaguiar96** [comment]
** left camera **

```
header:
  seq: 455
  stamp:
    secs: 0
    nsecs:         0
  frame_id: "zed_left_camera_optical_frame"
height: 720
width: 1280
distortion_model: "plumb_bob"
D: [0.0, 0.0, 0.0, 0.0, 0.0]
K: [692.7041625976562, 0.0, 636.4127807617188, 0.0, 692.7041625976562, 382.8067626953125, 0.0, 0.0, 1.0]
R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
P: [692.7041625976562, 0.0, 636.4127807617188, 0.0, 0.0, 692.7041625976562, 382.8067626953125, 0.0, 0.0, 0.0, 1.0, 0.0]
binning_x: 0
binning_y: 0
roi:
  x_offset: 0
  y_offset: 0
  height: 0
  width: 0
  do_rectify: False
```

** right camera **

```
header:
  seq: 88
  stamp:
    secs: 0
    nsecs:         0
  frame_id: "zed_right_camera_optical_frame"
height: 720
width: 1280
distortion_model: "plumb_bob"
D: [0.0, 0.0, 0.0, 0.0, 0.0]
K: [692.7041625976562, 0.0, 636.4127807617188, 0.0, 692.7041625976562, 382.8067626953125, 0.0, 0.0, 1.0]
R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
P: [692.7041625976562, 0.0, 636.4127807617188, -83.12450408935547, 0.0, 692.7041625976562, 382.8067626953125, 0.0, 0.0, 0.0, 1.0, 0.0]
binning_x: 0
binning_y: 0
roi:
  x_offset: 0
  y_offset: 0
  height: 0
  width: 0
  do_rectify: False
```

**Comment by miguelriemoliveira** [comment]
You could calibrate the intrinsics of both cameras using the ros tool

http://wiki.ros.org/camera_calibration

If you agree this should be added to:
https://github.com/lardemua/atom/issues/157

**Comment by eupedrosa** [comment]
They are almost the same... which is strange. And have no distortion, meaning the SDK outputs an undistorted image.

But there is a strange value in the right camera..
> P: [692.7041625976562, 0.0, 636.4127807617188, -83.12450408935547, 0.0, 692.7041625976562, 382.8067626953125, 0.0, 0.0, 0.0, 1.0, 0.0]

that -83 is strange !

**Comment by aaguiar96** [comment]
> that -83 is strange !

I noticed that too...



> They are almost the same... which is strange. And have no distortion, meaning the SDK outputs an undistorted image.

We have these topics for each camera:

`/zed/left/image_rect_color` : Color rectified left image.
`/zed/left/image_raw_color` : Color unrectified left image.
`/zed/left/camera_info` : Left camera calibration data.

**Comment by aaguiar96** [comment]
> If you agree this should be added to:
> #157

Sure @miguelriemoliveira, I can do that.
If we agree that this is the problem, I can go to the lab on Monday.

**Comment by eupedrosa** [comment]
I think that @miguelriemoliveira created a script to do camera calibration from the dataset, right? You can use that to calibrate it offline.

**Comment by aaguiar96** [comment]
> I think that @miguelriemoliveira created a script to do camera calibration from the dataset, right? You can use that to calibrate it offline.

Does the calibration work with charuco?

**Comment by eupedrosa** [comment]
Yes.

**Comment by aaguiar96** [comment]
> Yes.

Nice, I'll try that then!
@miguelriemoliveira can you tell me what is the script?

**Comment by miguelriemoliveira** [comment]
> I think that @miguelriemoliveira created a script to do camera calibration from the dataset, right? You can use that to calibrate it offline.

I don't remember this ... @eupedrosa I really have no memory of this ... do you know where it is?

You can try to calibrate from the bagfile.

**Comment by eupedrosa** [comment]
Quick fix: https://github.com/lardemua/atom/blob/3c501cacffc899da24329c122f9e78d4a3145fd3/atom_core/src/atom_core/objective_function.py#L152

Use the `K` instead of `P `

**Comment by eupedrosa** [comment]
> I don't remember this ... @eupedrosa I really have no memory of this ... do you know where it is?

https://github.com/lardemua/atom/blob/master/atom_calibration/scripts/convert_from_rwhe_dataset.py

The code is there ....

**Comment by miguelriemoliveira** [comment]
OK, thanks.

**Comment by miguelriemoliveira** [comment]
Tried a camera-velodyne optimization. This was the result:

![image](https://user-images.githubusercontent.com/11350998/84523885-fca3f500-acd0-11ea-8cb2-fe5fb94aacd3.png)


Something is still wrong. Will work on #178 so ease the debugging for this...

**Comment by aaguiar96** [comment]
Which collection did you use @miguelriemoliveira ?

It worked for me in collections without partial detections, such as collection 1 and 2 ... It should work for you too.

**Comment by miguelriemoliveira** [comment]
What do you mean it worked? Do you have a calibration that puts the lidar in a realistic location? Can you post an image?

**Comment by miguelriemoliveira** [comment]
This is what I get for collections 1 and 2, makes sense?

![image](https://user-images.githubusercontent.com/11350998/84532977-26641880-acdf-11ea-8678-e7f2840926a5.png)

If you had to guess, what is the distance you would say the lidar is to the camera?

I get this:

```
rosrun tf tf_echo c0_vlp16_frame c0_zed_camera_center
At time 1591985030.630
- Translation: [0.059, 0.049, -0.190]
- Rotation: in Quaternion [-0.009, 0.028, 0.014, 0.999]
            in RPY (radian) [-0.017, 0.057, 0.028]
            in RPY (degree) [-1.000, 3.268, 1.598]

```

**Comment by eupedrosa** [comment]
@miguelriemoliveira, did you use the `P` matrix in the optimization?

**Comment by aaguiar96** [comment]
> What do you mean it worked? Do you have a calibration that puts the lidar in a realistic location? Can you post an image?

No, I meant that the lidar points and the pattern point are aligned. That means that the objective function is working right? You also obtained that right?



> If you had to guess, what is the distance you would say the lidar is to the camera?

I would say 15/20cm. I think the sensor are aligned. Also, the camera has an inclination of ~11 degrees.

**Comment by aaguiar96** [comment]
> @miguelriemoliveira, did you use the `P` matrix in the optimization?

@eupedrosa does that make any difference? The error should be in the pattern location estimation, right? Or the objective function should also be able to correct the pattern location if we have a bad initial guess from `solvePnP`?

**Comment by eupedrosa** [comment]
The `K` matrix did not have that strange value.

**Comment by miguelriemoliveira** [comment]
Yes we do use it. The original idea was to use the K matrix, but I remember the atlascar2 bagfile had a problem with the K and P matrices being different, and we ended up using the P matrix.

I will commit a version that uses the K matrix.

**Comment by aaguiar96** [comment]
So, the next step is to recalibrate the camera intrinsics right?

I talked with a co-worker and the calibration that we're using is the one coming from the factory.
The problem is the estimation of the pattern position correct?

**Comment by miguelriemoliveira** [comment]
Since this is a very complex problem, I agree with @eupedrosa in that we should simplify whatever we have control of. Having a good intrinsics first guess (by using an intrinsics calibration that we know of) is definitely a good practice.

So I agree with @eupedrosa , we should do an intrinsics calibration.

But I am a bit lost right now: can you describe the problem that we have now? Because you said the velodyne is 15 cm above the camera, and the result I gave above is 19 cm. So I am a bit lost as to what is the problem ...

**Comment by aaguiar96** [comment]
> But I am a bit lost right now: can you describe the problem that we have now? Because you said the velodyne is 15 cm above the camera, and the result I gave above is 19 cm. So I am a bit lost as to what is the problem ...

I think since the objective function is working (the source and target points are aligned) the problem is on the estimation of the pattern location. The two sensors are aligned, and the solution is given the velodyne on the right of the ZED, which is wrong...

It should be something like this

![system](https://user-images.githubusercontent.com/35901587/84567289-8904f500-ad6f-11ea-9195-88eec2e03d4d.png)

**Comment by miguelriemoliveira** [comment]
Ok, but, just to be clear, now we don't seem to be having catastrophic errors. Rather, we are withing centimeters of what we think is the optimum solution right?

I am asking because this changes the tone of reasoning. It suggest we don't have a bug anymore, but should improve the accuracy of the procedures whenever possible.

Within this framework, precise intrinsic calibration is mandatory.

I suggest we close this issue (which is uge) and start a new one called "refine velodyne cost function" or something.

**Comment by aaguiar96** [comment]
> I am asking because this changes the tone of reasoning. It suggest we don't have a bug anymore, but should improve the accuracy of the procedures whenever possible.

Yes @miguelriemoliveira 
We do not have any bug anymore... I think we now should work on the intrinsics and a new dataset right?

> Within this framework, precise intrinsic calibration is mandatory.

Here's the problem, I don't think we have it right now...

> I suggest we close this issue (which is uge) and start a new one called "refine velodyne cost function" or something.

Agree, closing this one.
